File: lexer/lexer.go
package lexer

import (
	"unicode/utf8"
	"os"
	"fmt"
)


func Tokenize(chars []string) []Token {

	var tokens []Token
	ip := 0
	for ip <= len(chars)-1 {

		fmt.Println(chars[ip])
		if (ip == len(chars)){
			break
		}

		if (IsOneCharToken(chars[ip])){
			
			/* Check For OneCharTokens */
			if chars[ip] == "(" { tokens = BuildToken(tokens, chars[ip], LParen); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == ")" { tokens = BuildToken(tokens, chars[ip], RParen); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == "{" { tokens = BuildToken(tokens, chars[ip], LBrace); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == "}" { tokens = BuildToken(tokens, chars[ip], RBrace); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == "=" { tokens = BuildToken(tokens, chars[ip], Equals); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == "+" { tokens = BuildToken(tokens, chars[ip], Operator); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == "-" { tokens = BuildToken(tokens, chars[ip], Operator); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == "*" { tokens = BuildToken(tokens, chars[ip], Operator); if (ip+1>(len(chars)-1)){break}else{ip++} }
			if chars[ip] == "/" { tokens = BuildToken(tokens, chars[ip], Operator); if (ip+1>(len(chars)-1)){break}else{ip++} }
		
		}else {

			if IsNumerical(chars[ip]){
				Number := ""
				for IsNumerical(chars[ip]){
					Number += chars[ip]
					if (ip+1>(len(chars)-1)){break}else{ip++}

				}

				tokens = BuildToken(tokens, Number, Numeral)
			}

			if IsAlphabetical(chars[ip]){
				
				String := ""
				for IsAlphabetical(chars[ip]){
					String += chars[ip]					
					if (ip+1>(len(chars)-1)){break}else{ip++}
				}
				tf, k := IsKeyword(String)
				if tf {
					tokens = BuildToken(tokens, String, k)
				}else{
					tokens = BuildToken(tokens, String, Identifier)
				}

			}

			if IsSkippable(chars[ip]){
				if (ip+1>(len(chars)-1)){break}else{ip++}
			}

			if !IsAlphabetical(chars[ip]) == !IsNumerical(chars[ip]) == !IsOneCharToken(chars[ip]) == !IsSkippable(chars[ip]){
				f, _ := utf8.DecodeRuneInString(chars[ip])
				fmt.Println("Char That Cannot Be Handeled found in src ->", f)
				os.Exit(2)
			}

		}

	}
	
	return tokens

}

File: lexer/lexer.cs
using System;
using System.Collections.Generic;
using System.Text;
namespace Lexer
{
    public class lexer{

        public static List<Tokens.Token> Tokenize(string[] chars){
            
            List<Tokens.Token> tokens = new List<Tokens.Token>();
            int ip = 0;
            while (ip <= chars.Length-1){

                if (Tokens.IsOneCharToken(chars[ip])){

                    if (chars[ip] == "("){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.LParen);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == ")"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.RParen);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "{"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.LBrace);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "}"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.RBrace);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "<"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.LArrow);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == ">"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.RArrow);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "["){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.LBracket);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "]"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.RBracket);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "="){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.Equals);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "+"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.Operator);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "-"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.Operator);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "*"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.Operator);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    if (chars[ip] == "/"){tokens=Tokens.BuildToken(tokens, chars[ip], Tokens.TokenTypes.Operator);if (ip+1>(chars.Length-1)){break;}else{ip++;}}
                    
                }else{

                    if (Tokens.IsNumerical(chars[ip])){

                        String Number = "";
                        while (Tokens.IsNumerical(chars[ip])){
                            
                            Number += chars[ip];
                            if (ip+1>(chars.Length-1)){break;}else{ip++;}

                        }
                        tokens = Tokens.BuildToken(tokens, Number, Tokens.TokenTypes.Numeral);

                    }

                    if (Tokens.IsAlphabetical(chars[ip])){

                        string String = "";
                        while (Tokens.IsAlphabetical(chars[ip])){
                            String += chars[ip];
                            if (ip+1>(chars.Length-1)){break;}else{ip++;}
                        }
                        bool tf = Tokens.IsKeyword(String)[0];
                        Tokens.TokenTypes k = Tokens.IsKeyword(String)[1];
                        if (tf){
                            tokens = Tokens.BuildToken(tokens, String, k);
                        }else{
                            tokens = Tokens.BuildToken(tokens, String, Tokens.TokenTypes.Identifier);
                        }

                    }


                    if (Tokens.IsSkippable(chars[ip])){
                        if (ip+1>(chars.Length-1)){break;}else{ip++;}
                    }

                    if (!Tokens.IsAlphabetical(chars[ip]) == !Tokens.IsNumerical(chars[ip]) == !Tokens.IsOneCharToken(chars[ip]) == !Tokens.IsSkippable(chars[ip])){
                        Console.WriteLine("Char That Cannot Be Handeled found in src -> { "+ BitConverter.ToString(Encoding.UTF8.GetBytes("e"))+ " }");
                        Environment.Exit(2);
                    }

                }

                // if (ip == chars.Length-1){
                //     break;
                // }

            }

            tokens = Tokens.BuildToken(tokens, "EndOfFile", Tokens.TokenTypes.EOF);

            return tokens;

        }

    }
    
}